nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/signer/AwsCustomSignerFactory.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.aws.signer;
+
+import com.amazonaws.auth.Signer;
+
+public interface AwsCustomSignerFactory {
```

Comment:
This interface effectively introduces a new internal extension point, which seems like it could become difficult to maintain.

Reply:
@exceptionfactory Thanks for your hint about the `sign()` method parameters!

Request URL is available on `SignableRequest` as you mentioned.

For region and service type we need to type cast `SignableRequest` to `DefaultRequest`:
```
((DefaultRequest) request).getHandlerContext(HandlerContextKey.SIGNING_REGION)
((DefaultRequest) request).getHandlerContext(HandlerContextKey.SERVICE_ID)
```
As `DefaultRequest` is the only implementation of the interface (available in the dependencies we use), the type cast seems to be safe to use in custom signer implementations.

Reply:
@exceptionfactory Updated the PR.
Found a more straightforward way to access region and service type: `AWS4Signer` implementation of the `Signer` interface has `serviceName` and `regionName` fields and this class may be used as a base class for custom signers.